pkg/agent/bgp/engine/gobgp.go

```diff
@@ -0,0 +1,295 @@
+		if peer == nil {
+			return
+		}
+		peerStatus := PeerStatus{}
+		if peer.Transport != nil {
+			peerStatus.Port = int32(peer.GetTransport().GetRemotePort())
+		}
+		if peer.Conf != nil {
+			peerStatus.Address = peer.GetConf().GetNeighborAddress()
+			peerStatus.ASN = int32(peer.GetConf().GetPeerAsn())
+		}
+		if peer.GetEbgpMultihop() != nil && peer.GetEbgpMultihop().GetEnabled() {
+			peerStatus.MultihopTTL = int32(peer.GetEbgpMultihop().GetMultihopTtl())
+		} else {
+			peerStatus.MultihopTTL = DefaultBGPMultihopTTL
+		}
+		if peer.GetGracefulRestart() != nil {
+			peerStatus.GracefulRestartTimeSeconds = int32(peer.GetGracefulRestart().GetRestartTime())
+		}
+		if peer.State != nil {
+			peerStatus.SessionState = toSessionState(peer.GetState().GetSessionState())
+			if peerStatus.SessionState == SessionEstablished && peer.GetTimers() != nil && peer.GetTimers().GetState() != nil {
+				peerStatus.UptimeSeconds = int(time.Since(peer.GetTimers().GetState().GetUptime().AsTime()).Seconds())
+			}
+		}
+		peerStatuses = append(peerStatuses, peerStatus)
+	}
+
+	request := &gobgpapi.ListPeerRequest{EnableAdvertised: true}
+	err := g.server.ListPeer(context.TODO(), request, fn)
+	if err != nil {
+		return peerStatuses, err
+	}
+	return peerStatuses, nil
+}
+
+func (g *GoBGPServer) AdvertiseRoutes(routes []Route, ipFamily net.IPFamily) error {
+	for i := range routes {
+		request := &gobgpapi.AddPathRequest{Path: toGoBGPPath(&routes[i], ipFamily)}
+		if _, err := g.server.AddPath(context.TODO(), request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *GoBGPServer) WithdrawRoutes(routes []Route, ipFamily net.IPFamily) error {
+	for i := range routes {
+		request := &gobgpapi.DeletePathRequest{Path: toGoBGPPath(&routes[i], ipFamily)}
+		if err := g.server.DeletePath(context.TODO(), request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *GoBGPServer) GetRoutes(routeType RouteType, peerAddress string, ipFamily net.IPFamily) ([]Route, error) {
+	var gobgpTableType gobgpapi.TableType
+	if routeType == RouteAdvertised {
+		gobgpTableType = gobgpapi.TableType_ADJ_OUT
+	} else if routeType == RouteReceived {
+		gobgpTableType = gobgpapi.TableType_ADJ_IN
+	} else {
+		return nil, fmt.Errorf("invalid route type: %v", routeType)
+	}
+	if peerAddress == "" {
+		return nil, fmt.Errorf("BGP peer IP address must be specified")
+	}
+
+	var routes []Route
+	fn := func(destination *gobgpapi.Destination) {
+		if destination == nil {
+			return
+		}
+		routes = append(routes, Route{Prefix: destination.GetPrefix()})
+	}
+
+	request := &gobgpapi.ListPathRequest{
+		TableType: gobgpTableType,
+		Family:    &gobgpapi.Family{Afi: toGoBGPFamilyAfi(ipFamily), Safi: gobgpapi.Family_SAFI_UNICAST},
+		Name:      peerAddress,
+	}
+
+	if err := g.server.ListPath(context.TODO(), request, fn); err != nil {
+		return nil, err
+	}
+	return routes, nil
+}
+
+func toGoBGPPath(route *Route, ipFamily net.IPFamily) *gobgpapi.Path {
```

Comment:
the ipFamily parameter is redundant here. It will need to parse the prefix anyway, `prefix.Addr().Is4()` can be used.

Reply:
Will refine this.